### PR TITLE
show tos and privacy links on oauth screen

### PR DESCRIPTION
**Pull request by cpapazian**
https://github.com/18F/myusa/issues/445

### CODE REVIEW

**Comment by yozlet** [comment]
Good so far, but missing this item from [Joe's spec](https://github.com/18F/myusa/issues/445#issuecomment-71943253):
* In the validation rules, check that each link exists if it is entered (HTTP 200)

File: lib/doorkeeper_patches/models/application.rb [hunk 1]
```diff
@@ -11,16 +11,22 @@ class Doorkeeper::Application
...
-                                                      tos_link: ->(values) { Rails.application.routes.url_helpers.legal_path(anchor: 'terms-of-service') }
-
-  validates_presence_of :organization, if: :federal_agency,
-                                       accept: true,
-                                       allow_nil: false,
-                                       message: :federal_agency_org_required
+  validates_acceptance_of :terms_of_service_accepted,
+    if: :federal_agency, accept: true, allow_nil: false,
+    message: :federal_agency_tos_required,
+    tos_link: ->(values) { Rails.application.routes.url_helpers.legal_path(anchor: 'terms-of-service') }
```
Comment (yozlet) [inline lib/doorkeeper_patches/models/application.rb]: Sorry for my ignorance - what does this line do?

Comment (cpapazian) [inline lib/doorkeeper_patches/models/application.rb]: just changing the spacing here, but ...

http://apidock.com/rails/ActiveModel/Validations/HelperMethods/validates_acceptance_of

validates that boolean field is true. the `if` argument means that we only require it in certain conditions (in this case, when the `federal_agency` method is true). `message` points to a custom error message, and `tos_link` is an arbitrary argument ... in this case, it's used by the error message.

Comment (yozlet) [inline lib/doorkeeper_patches/models/application.rb]: No, I get that bit, I'm talking about `Rails.application.routes.url_helpers.legal_path(anchor: 'terms-of-service')`

Comment (cpapazian) [inline lib/doorkeeper_patches/models/application.rb]: if you look in `en.yml` for `federal_agency_tos_required`, you'll notice it has a `tos_link` attribute. the lambda in the diff above will call `legal_path(anchor: 'terms-of-service')` get evaluated at run time to `/legal#terms-of-service`, and then the active record error message code will call `I18n.t(:message, tos_link: ...)`

**Comment by adelevie** [comment]
@yozlet, ahoy!

> In the validation rules, check that each link exists if it is entered (HTTP 200)

We still need this?

**Comment by adelevie** [comment]
Locally, I'm getting a lot of failing specs. I have a fresh dev setup, so there's probably something simple I need to add.

The spec failures stem from

```
     NoMethodError:
       undefined method `tos_link' for #<Doorkeeper::Application:0x007fe3174abf30>
     # ./spec/requests/v1/api_spec.rb:4:in `block (2 levels) in <top (required)>'
     # ./spec/requests/v1/api_spec.rb:253:in `block (5 levels) in <top (required)>'
```

**Comment by cpapazian** [comment]
>In the validation rules, check that each link exists if it is entered (HTTP 200)

i didn't implement that, but forgot to mention it here. i don't think it's needed. i validate the URL, which is all we do for other client application URLs.

**Comment by adelevie** [comment]
Looks good to me. Thanks @cpapazian for helping run this locally.

**Comment by adelevie** [comment]
Wrote this up, too: http://stackoverflow.com/questions/29156975/programmatically-detect-and-catch-infinite-redirect-loops-in-rails

**Pull request merged**